src/ItemGrid.cpp

```diff
@@ -482,26 +490,28 @@ cItem cItemGrid::RemoveOneItem(int a_SlotNum)
 	if (!IsValidSlotNum(a_SlotNum))
 	{
 		LOGWARNING("%s: Invalid slot number %d out of %d slots, ignoring the call, returning empty item",
-			__FUNCTION__, a_SlotNum, m_NumSlots
+			__FUNCTION__, a_SlotNum, m_Slots.size()
 		);
 		return cItem();
 	}
 
 	// If the slot is empty, return an empty item
-	if (m_Slots[a_SlotNum].IsEmpty())
+	if (m_Slots.GetAt(a_SlotNum).IsEmpty())
```

Comment:
> m_Slots.data()[x] is the same except saving the result of data() can be used to avoid duplicating nullptr checks.

Sorry, but either prove me this optimisation with a profiler or just make the code readable! Our processors all know about [branch prediction](https://en.wikipedia.org/wiki/Branch_predictor) and this is a good example, since we either have a lot of "not taken", when we only access data or we have a lot of "taken", when the array is initialized.
Furthermore I also have a technique to eliminate the nullptr checks:
```cpp
static std::array<MAX_CONTAINER_SIZE, Item> emptyContainer;
class Lazy {
Item *first = &emptyContainer[0]
const operator[](int i) { return first[í]; } // NO nullptr checks necessary!!!
```

Sorry, but I remember myself thinking about optimizing a A*-Algorithm. I first thought about using pointers instead of the real objects in the `priority_queue`(to make swapping faster). But after I've looked at the system monitor, I've noticed, that the program ate the whole RAM and the system moved parts of the program data on the swap. When I realized this, I knew that my optimisation woudn't provide any benefits...

> In non-const contexts, m_Slots[x] will cause allocations because the reference could be used to mutate the container.

Jep, I know. But `IsEmpty` should be const, so m_Slots[x] shouldn't cause an allocation. The whole point of the `cLazyArray` is to hide the fact of lazy storage, where we can. Only in case of things like `m_Slots[x].DamageItem`, I would handle the allocation issues via `IsStorageAllocated`. But most functions only access/search in the array for an item, so I would only have a **few** places of `IsStorageAllocated` 

> I admit it's a bit of a funny dance. Maybe I need to expand on the doxy comments for cLazyArray to clarify what might allocate?

No, I think comments are not the best thing. I would only comment the **few** fast bailouts via `IsStorageAllocated`. Either use the current design of `operator[]` and abstract the laizy initialization away from the user (which **might** cause some unintended initialisations). Or make `operator[]` always return `const cItem` and make an additional function `Change(int)`, which will allocate the storage. Then you will be explicit, by saying `m_Items.Change(x).DamageItem` and `m_Items[x].IsEmpty()`.

Reply:
> operator[] has the same behaves as any other container

I think, you mean the signature. Because it has two different behaviours, based on the constness.

> I'm not convinced that having both operator[] and Change be special cases is any clearer.

The question is: Should it be clear to the reader, when the container will allocate space and when not?

If it should be clear, then make `operator[]` consistent (either always allocate space or never allocate space) and implement a second function for the other behaviour (`GetAt` or `Change`). Then the reader always know, that if for example GetAt is called, no storage will be allocated and every other `m_Item[x]` will allocate space.

If the user shouldn't need to care, if the class allocates space or not, then always use `operator[]` and maybe leave the `IsStorageAllocated` bail out at areas, where it would allocate storage, but you can do without it. But then you should add a comment, so that the user knows, that the bail out is used to save an allocation.

Reply:
>use the specialisation GetAt all the time, when you doesn't want to allocate space

That is already the intended use, the fact that `operator[] const` doesn't need to allocate is a happy accident.
Look at `cItemGrid::ChangeSlotCount` for example, at first it uses `GetAt` but after `IsEmpty` is false, you know the array has already allocated and there's no need to avoid using `[]` to modify the array.

To add allocation to `operator [] const`, I think I would need a more compelling argument than just being identical to the non-`const` version.

Reply:
> first it uses GetAt but after IsEmpty is false, you know the array has already allocated and there's no need to avoid using [] to modify the array.

Sorry, but I don't agree. Code changes/gets refactored. Maybe somebody has the idea to first calculate the resulting count maybe like `std::min(std::max(0, m_Slots[a_SlotNum].m_ItemCount + a_AddToCount), Handler->GetMaxStackSize()))` for some reason. Now he just made the mistake of always allocating the space.

Furthermore it may be confusing to the reader: Does `GetAt` return something different than `operator[]`? Or when should I now use `GetAt`, `operator[]` or `IsStorageAllocated`?

Why not just call `GetAt` at every place, where you are okay with a constant object? This would make it consistent and remove the usage of some unnecessary [IsStorageAllocated calls](https://github.com/cuberite/cuberite/pull/4083/files#diff-3e8894a3ba7bf8753efec6461864b809R409).

> To add allocation to operator [] const, I think I would need a more compelling argument than just being identical to the non-const version.

Here comes the better argument 👊  : Just delete it 🗑 . Then the behaviour is consistent, the `const_cast` trick is not necessary (and we just saved some bytes).

Reply:
>it may be confusing to the reader

I don't really see how constantly alternating between one and the other would make it any less confusing. For example, RemoveItem would have:
```cpp
if (m_Slots.GetAt(i).IsEqual(a_ItemStack))
{
	int NumToRemove = [...];
	m_Slots[i].m_ItemCount -= NumToRemove;
```
I don't think I could tell the difference between `GetAt(i)` and `[i]` without reading the doxy/code.

>remove the usage of some unnecessary IsStorageAllocated calls.

When it's turning an O(n) operation into O(1) I don't really see it as a waste.  I know it's not realistically going to effect performance but it certainly doesn't hurt.

Reply:
> I don't think I could tell the difference between GetAt(i) and [i] without reading the doxy/code.

`GetAt` returns a non modifiable object (const), while `[i]` returns a modifiable object. The method signature is enough to show the difference. And you could put in the comments of `[i]`, that `GetAt` should be preferred, when possible.

>  I know it's not realistically going to effect performance but it certainly doesn't hurt.

First of all it hurts readability. I now have to look up one more function and understand, that it is equivalent to the for loop with an empty list. Furthermore this just destroys the abstraction of the `cLazyArray` class, because I now need to look up, if the extra condition is necessary (maybe it prevents accessing uninitialized space).

And the extra point is, that this condition currently changes the behaviour of some functions. Look at the `Clear` function:

```cpp

void cItemGrid::Clear(void)
{
	if (!m_Slots.IsStorageAllocated())
	{
		return;  // Already clear
	}

	for (int i = 0; i < m_Slots.size(); i++)
	{
		m_Slots[i].Empty();
		TriggerListeners(i);
	}
}
```
The new behaviour is, that if no storage is allocated, we don't call any listeners. But on the other hand, if the storage is allocated (**which doesn't imply, that the container contains items**) we call `TriggerListener` for every inventory slot.

For me the increased readability effort and the different behaviour hurt much more, than the 0,0001% performance gain. I always prefer code, that is easy to read over some subtile performance improvements. And if code gets more compilcated due to performance or some overlooked behaviour, it should [be commented](https://en.wikipedia.org/wiki/Lions%27_Commentary_on_UNIX_6th_Edition,_with_Source_Code#%22You_are_not_expected_to_understand_this%22).

P.S.: Would you change the `find`-Method of the STL-vector to something like this (It finds elements at the end with O(1) 😈 )?:

```cpp
Iterator find(const Elem & toFind) const
{
    if(toFind == *this[size() - 1])
       return end() - 1;

    auto it = begin();
    for(; it < end(); ++it)
        if(toFind == *it)
            break;
    return it;
}
```

Reply:
>The method signature is enough to show the difference.

It doesn't tell me that `GetAt` won't cause allocation which is what actually matters.

>this just destroys the abstraction of the `cLazyArray` class

Maybe we disagree on where responsibility lies here. This is what I think:
`cItemGrid` is the one that abstracts away the fact that the storage might not be allocated. `cLazyArray` is about _embracing_ that. You wouldn't ask for `std::shared_ptr` to hide that it's reference counted or `std::list` to hide that it's a doubly-linked list.  It's not an abstraction failure; it's the whole point of the class.


As for `cItemGrid::Clear`: not calling listeners when empty cells are re-emptied is already happening in `cItemGrid::EmptySlot`.  So listeners have to be tolerant to this behaviour already.

Regarding your `find`: it's not really the same because the work to check `IsStorageAllocated` is always being done anyway.  As a counter-point consider this implementation of find:
```cpp
It find(const Elem & toFind) const
{
    auto ret = end();
    for(auto it = begin(); it < end(); ++it)
        if(toFind == *it)
            ret = it;
    return ret;
}
```
Where even after finding the item it continues to loop over the container anyway.  I see the obviously wasted work and it just seems "immoral".

(Ignoring that both implementations of find give the wrong results when multiple elements match)